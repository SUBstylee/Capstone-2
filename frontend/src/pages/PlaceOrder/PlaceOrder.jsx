import { useEffect } from "react";
import { useNavigate,Link } from "react-router-dom";
import {Button,Row,Col,ListGroup,Image,Card} from 'react-bootstrap';
import { useDispatch,useSelector } from "react-redux";
import Message from "../../components/Message/Message";
import CheckoutSteps from "../../components/CheckoutSteps/CheckoutSteps";
import { createOrder } from "../../actions/orderActions";
import MetaWrapper from "../../components/MetaWrapper/MetaWrapper";

const PlaceOrder = () => {
    const dispatch=useDispatch();
    const navigate=useNavigate();
    const cart=useSelector(state=>state.cart);
    //calc prices
    cart.orderItems=cart.cartItems.reduce((acc,item)=>acc+item.price*item.qty,0);
    cart.shippingPrice=cart.orderItems>=100?0.00:10.00;
    cart.taxPrice=Number(0.12*cart.orderItems);
    cart.totalPrice=(Number(cart.orderItems)+Number(cart.shippingPrice)+cart.taxPrice);

    const orderCreate=useSelector(state=>state.orderCreate);
    const {order,success,error}=orderCreate;

    useEffect(()=>{
        if(success){
            navigate(`/order/${order._id}`);
        };
        // eslint-disable-next-line
    },[navigate,success]);

    const placeOrderHandler=()=>{
        dispatch(createOrder({
            orderItems: cart.cartItems,
            shippingAddress: cart.shippingAddress,
            paymentMethod: cart.paymentMethod,
            itemsPrice: cart.itemsPrice,
            shippingPrice: cart.shippingPrice,
            taxPrice: Number(cart.taxPrice).toFixed(2),
            totalPrice: Number(cart.totalPrice).toFixed(2),
        }));
    };

    return (
        <div className="PlaceOrder">
            <MetaWrapper title='TAA-Place Order'/>
            <CheckoutSteps step1 step2 step3 step4 />
            <Row>
                <Col md={8}>
                    <ListGroup variant='flush'>
                        <ListGroup.Item>
                            <h2>Shipping</h2>
                            <p>
                                <strong>Address: </strong>
                                {cart.shippingAddress.address},{' '}
                                {cart.shippingAddress.city},{' '}
                                {cart.shippingAddress.postalCode},{' '}
                                {cart.shippingAddress.country}
                            </p>
                        </ListGroup.Item>
                        <ListGroup.Item>
                            <h2>Payment Method</h2>
                            <strong>Method: </strong>
                            {cart.paymentMethod}  
                        </ListGroup.Item>  
                        <ListGroup.Item>
                            <h2>Items In Order</h2>
                            {cart.cartItems.length===0? <Message>Your cart is empty...</Message>:(
                                <ListGroup variant="flush">
                                    {cart.cartItems.map((item,index)=>(
                                        <ListGroup.Item key={index}>
                                            <Row>
                                                <Col md={1}>
                                                    <Image src={item.image} alt={item.name} fluid rounded />
                                                </Col>    
                                                <Col>
                                                    <Link to={`/products/${item.product}`}>{item.name}</Link>
                                                </Col>
                                                <Col md={4}>
                                                    {item.qty} X ${item.price} = <strong>${item.qty * item.price}</strong>
                                                </Col>
                                            </Row>    
                                        </ListGroup.Item>
                                    ))}    
                                </ListGroup>
                            )}
                        </ListGroup.Item>  
                    </ListGroup>    
                </Col>
                <Col md={4}>
                    <Card>
                        <ListGroup variant='flush'>
                            <ListGroup.Item>
                                <h2>Order Summary</h2>
                            </ListGroup.Item>
                            <ListGroup.Item>
                                <Row>
                                    <Col>Subtotal</Col>
                                    <Col>${cart.orderItems.toFixed(2)}</Col>
                                </Row>
                            </ListGroup.Item>
                            <ListGroup.Item>
                                <Row>
                                    <Col>Shipping</Col>
                                    <Col>${cart.shippingPrice.toFixed(2)}</Col>
                                </Row>
                            </ListGroup.Item>
                            <ListGroup.Item>
                                <Row>
                                    <Col>Tax</Col>
                                    <Col>${cart.taxPrice.toFixed(2)}</Col>
                                </Row>
                            </ListGroup.Item>
                            <ListGroup.Item>
                                <Row>
                                    <Col>Total</Col>
                                    <Col><strong>${cart.totalPrice.toFixed(2)}</strong></Col>
                                </Row>
                            </ListGroup.Item>
                            <ListGroup.Item>
                                {error&&<Message variant='danger'>{error}</Message>}
                            </ListGroup.Item>
                            <ListGroup.Item>
                                <Button type='button' className="btn-block" disabled={cart.cartItems.length===0} onClick={placeOrderHandler}>{cart.cartItems.length===0?(<del>Place Order</del>):('Place Order')}</Button>
                            </ListGroup.Item>
                        </ListGroup>
                    </Card>
                </Col>
            </Row>  
        </div>
    );
};

export default PlaceOrder;