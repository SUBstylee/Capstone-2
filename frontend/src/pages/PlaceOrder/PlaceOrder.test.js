import { render } from '@testing-library/react';
import PlaceOrder from './PlaceOrder.jsx';
import store from '../../store';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

test("renders Home component without crashing", () => {
    const { getByText } = render(
        <Provider store={store}><BrowserRouter><PlaceOrder /></BrowserRouter></Provider>
    );

    expect(getByText("Items In Order")).toBeInTheDocument();
});

test("matches snapshot", function () {
    const { asFragment } = render(
        <Provider store={store}><BrowserRouter><PlaceOrder /></BrowserRouter></Provider>
    );
    expect(asFragment()).toMatchSnapshot();
});