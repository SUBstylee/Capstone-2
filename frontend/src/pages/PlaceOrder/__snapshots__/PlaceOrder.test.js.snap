// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="PlaceOrder"
  >
    <div
      class="CheckoutSteps"
    >
      <div
        class="justify-content-center mb-4 nav"
      >
        <div
          class="nav-item"
        >
          <a
            class="nav-link"
            data-rr-ui-event-key="/login"
            href="/login"
          >
            Sign In
          </a>
        </div>
        <div
          class="nav-item"
        >
          <a
            class="nav-link"
            data-rr-ui-event-key="/shipping"
            href="/shipping"
          >
            Shipping
          </a>
        </div>
        <div
          class="nav-item"
        >
          <a
            class="nav-link"
            data-rr-ui-event-key="/payment"
            href="/payment"
          >
            Payment
          </a>
        </div>
        <div
          class="nav-item"
        >
          <a
            class="nav-link"
            data-rr-ui-event-key="/placeorder"
            href="/placeorder"
          >
            Place Order
          </a>
        </div>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-md-8"
      >
        <div
          class="list-group list-group-flush"
        >
          <div
            class="list-group-item"
          >
            <h2>
              Shipping
            </h2>
            <p>
              <strong>
                Address: 
              </strong>
              , , , 
            </p>
          </div>
          <div
            class="list-group-item"
          >
            <h2>
              Payment Method
            </h2>
            <strong>
              Method: 
            </strong>
          </div>
          <div
            class="list-group-item"
          >
            <h2>
              Items In Order
            </h2>
            <div
              class="fade alert alert-info show"
              role="alert"
            >
              Your cart is empty...
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-md-4"
      >
        <div
          class="card"
        >
          <div
            class="list-group list-group-flush"
          >
            <div
              class="list-group-item"
            >
              <h2>
                Order Summary
              </h2>
            </div>
            <div
              class="list-group-item"
            >
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  Subtotal
                </div>
                <div
                  class="col"
                >
                  $0.00
                </div>
              </div>
            </div>
            <div
              class="list-group-item"
            >
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  Shipping
                </div>
                <div
                  class="col"
                >
                  $10.00
                </div>
              </div>
            </div>
            <div
              class="list-group-item"
            >
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  Tax
                </div>
                <div
                  class="col"
                >
                  $0.00
                </div>
              </div>
            </div>
            <div
              class="list-group-item"
            >
              <div
                class="row"
              >
                <div
                  class="col"
                >
                  Total
                </div>
                <div
                  class="col"
                >
                  <strong>
                    $10.00
                  </strong>
                </div>
              </div>
            </div>
            <div
              class="list-group-item"
            />
            <div
              class="list-group-item"
            >
              <button
                class="btn-block btn btn-primary"
                disabled=""
                type="button"
              >
                <del>
                  Place Order
                </del>
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
